engine/pkg/dm/message_agent.go

```diff
@@ -0,0 +1,404 @@
+	responseTp
+)
+
+// message use for asynchronous message and synchronous request/response.
+type message struct {
+	ID      messageID
+	Type    messageType
+	Command string
+	Payload interface{}
+}
+
+// messageIDAllocator is an id allocator for p2p message system
+type messageIDAllocator struct {
+	mu sync.Mutex
+	id messageID
+}
+
+// alloc allocs a new message id
+func (a *messageIDAllocator) alloc() messageID {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	a.id++
+	return a.id
+}
+
+// Sender defines an interface that supports send message
+type Sender interface {
+	SendMessage(ctx context.Context, topic p2p.Topic, message interface{}, nonblocking bool) error
+}
+
+// messagePair implement a simple synchronous request/response message pattern since the lib currently only support asynchronous message.
+// Caller should persist the request message if needed.
+// Caller should add retry mechanism if needed.
+type messagePair struct {
+	// messageID -> response channel
+	// TODO: limit the MaxPendingMessageCount if needed.
+	pendings    sync.Map
+	idAllocator *messageIDAllocator
+}
+
+// newMessagePair creates a new MessagePair instance
+func newMessagePair() *messagePair {
+	return &messagePair{
+		idAllocator: &messageIDAllocator{},
+	}
+}
+
+// sendRequest sends a request message and wait for response.
+func (m *messagePair) sendRequest(ctx context.Context, topic p2p.Topic, command string, req interface{}, sender Sender) (interface{}, error) {
+	msg := message{ID: m.idAllocator.alloc(), Type: requestTp, Command: command, Payload: req}
+	respCh := make(chan interface{}, 1)
+	m.pendings.Store(msg.ID, respCh)
+	defer m.pendings.Delete(msg.ID)
+
+	if err := sender.SendMessage(ctx, topic, msg, false /* nonblock */); err != nil {
+		return nil, err
+	}
+
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case resp := <-respCh:
+		return resp, nil
+	}
+}
+
+// sendResponse sends a response with message ID.
+func (m *messagePair) sendResponse(ctx context.Context, topic p2p.Topic, id messageID, command string, resp interface{}, sender Sender) error {
+	msg := message{ID: id, Type: responseTp, Command: command, Payload: resp}
+	return sender.SendMessage(ctx, topic, msg, false /* nonblock */)
+}
+
+// onResponse receives and pairs a response message.
+func (m *messagePair) onResponse(id messageID, resp interface{}) error {
+	respCh, ok := m.pendings.Load(id)
+	if !ok {
+		return errors.Errorf("request %d not found", id)
+	}
+
+	select {
+	case respCh.(chan interface{}) <- resp:
+		return nil
+	default:
+	}
+	return errors.Errorf("duplicated response of request %d, and the last response is not consumed", id)
+}
+
+// MessageAgent defines interface for message communication.
+type MessageAgent interface {
+	Init(ctx context.Context) error
+	Tick(ctx context.Context) error
+	Close(ctx context.Context) error
+	UpdateSender(senderID string, sender Sender) error
+	SendMessage(ctx context.Context, senderID string, command string, msg interface{}) error
```

Comment:
Do you think it is possible to make `SendMessage` and `SendRequest` generic to avoid using `interface{}`?